### ISSUE TITLE
Segfault when a combatant's attacks are removed in a combat event

### ISSUE THREAD

**Issue by elidupree** [wesnoth/wesnoth#5914]
### Game and System Information
Wesnoth 1.15.14 on both Debian and Windows 10.

### Describe the bug
Segfault when a combatant's attacks are removed in a combat event.

### To Reproduce
Download this savefile and perform a melee attack from the unit on 9,8 to the adjacent Wraith. (You might have to retry if the attacker misses.)
[2p_Hamlets-Auto-Save11.gz](https://github.com/wesnoth/wesnoth/files/6755874/2p_Hamlets-Auto-Save11.gz)

### Expected behavior
EoHS Lua code removes the defender's attacks in an `attacker hits` event. In previous versions of Wesnoth, this made the defender stop swinging, according to this note in my code:
```
    -- After a battle starts, it is literally impossible for one side's number of remaining strikes to be reduced without ending the combat entirely.
    -- The best we can do is to remove all their attacks, which essentially makes them swing a null attack with zero chance to hit.
    -- To make this look right, we also have to override their opponent's defense animation with an empty animation.
```

### Backtrace
From some quick investigation, it appears that `udisplay.cpp:624` assumes that the attack is not null, but `attacker_stats->weapon` already null on `attack.cpp:1094`.

```
#0  0x0000555555e3cebc in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count(std::__weak_count<(__gnu_cxx::_Lock_policy)2> const&) (this=0x7fffffff97a8, __r=...)
    at /usr/include/c++/8/bits/shared_ptr_base.h:893
#1  0x0000555555f74267 in std::__shared_ptr<attack_type const, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<attack_type, void>(std::__weak_ptr<attack_type, (__gnu_cxx::_Lock_policy)2> const&)
    (this=0x7fffffff97a0, __r=...) at /usr/include/c++/8/bits/shared_ptr_base.h:1191
#2  0x0000555555f72bf7 in std::shared_ptr<attack_type const>::shared_ptr<attack_type, void>(std::weak_ptr<attack_type> const&)
    (this=0x7fffffff97a0, __r=<error reading variable: Cannot access memory at address 0x8>) at /usr/include/c++/8/bits/shared_ptr.h:276
#3  0x0000555555f71e95 in std::enable_shared_from_this<attack_type>::shared_from_this() const (this=0x0) at /usr/include/c++/8/bits/shared_ptr.h:659
#4  0x000055555601f36a in unit_display::unit_attack(display*, game_board&, map_location const&, map_location const&, int, attack_type const&, std::shared_ptr<attack_type const>, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const*, bool)
    (disp=0x7fff90edff40, board=..., a=..., b=..., damage=0, attack=..., secondary_attack=
    std::shared_ptr<const attack_type> (use count 3, weak count 1) = {...}, swing=1, hit_text="", drain_amount=0, att_text="", extra_hit_sounds=0x7fffffff9c10, attacking=false)
    at src/units/udisplay.cpp:624
#5  0x00005555561226a7 in (anonymous namespace)::attack::perform_hit(bool, statistics::attack_context&) (this=0x7fffffffa1f0, attacker_turn=false, stats=...) at src/actions/attack.cpp:1094
#6  0x0000555556125522 in (anonymous namespace)::attack::perform() (this=0x7fffffffa1f0) at src/actions/attack.cpp:1440
#7  0x00005555561263e9 in attack_unit(map_location const&, map_location const&, int, int, bool) (attacker=..., defender=..., attack_with=0, defend_with=0, update_display=true)
    at src/actions/attack.cpp:1562
#8  0x000055555612647f in attack_unit_and_advance(map_location const&, map_location const&, int, int, bool) (attacker=..., defender=..., attack_with=0, defend_with=0, update_display=true)
    at src/actions/attack.cpp:1571
#9  0x0000555555f18565 in synced_command_func_attack(config const&, bool, bool, synced_command::error_handler_function) (child=..., show=true, error_handler=...)
    at src/synced_commands.cpp:227
#10 0x0000555555f1fb2f in synced_context::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, config const&, bool, bool, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>) (commandname="attack", data=..., use_undo=true, show=true, error_handler=...) at src/synced_context.cpp:67
#11 0x0000555555f1fea5 in synced_context::run_and_store(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, config const&, bool, bool, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>) (commandname="attack", data=..., use_undo=true, show=true, error_handler=...)
    at src/synced_context.cpp:95
#12 0x0000555555f1ffae in synced_context::run_and_throw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, config const&, bool, bool, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>) (commandname="attack", data=..., use_undo=true, show=true, error_handler=...)
    at src/synced_context.cpp:109
#13 0x0000555556465edf in events::mouse_handler::attack_enemy_(map_location const&, map_location const&, int) (this=0x7fffffffc740, att_loc=..., def_loc=..., choice=0)
    at src/mouse_events.cpp:1358
#14 0x0000555556465791 in events::mouse_handler::attack_enemy(map_location const&, map_location const&, int) (this=0x7fffffffc740, attacker_loc=..., defender_loc=..., choice=0)
    at src/mouse_events.cpp:1297
#15 0x0000555556462af8 in events::mouse_handler::move_action(bool) (this=0x7fffffffc740, browse=false) at src/mouse_events.cpp:884
#16 0x0000555556462572 in events::mouse_handler::select_or_action(bool) (this=0x7fffffffc740, browse=false) at src/mouse_events.cpp:815
#17 0x0000555556431e74 in play_controller::hotkey_handler::select_and_action() (this=0x555563111e40) at src/hotkey/hotkey_handler.cpp:122
#18 0x00005555568c62d3 in hotkey::command_executor::do_execute_command(hotkey::hotkey_command const&, int, bool, bool) (this=0x555563111e40, cmd=..., press=true, release=false)
    at src/hotkey/command_executor.cpp:336
#19 0x0000555556432979 in play_controller::hotkey_handler::do_execute_command(hotkey::hotkey_command const&, int, bool, bool)
    (this=0x555563111e40, cmd=..., index=-1, press=true, release=false) at src/hotkey/hotkey_handler.cpp:275
#20 0x00005555568c7b5a in hotkey::command_executor::execute_command_wrap(hotkey::command_executor::queued_command const&) (this=0x555563111e40, command=...)
    at src/hotkey/command_executor.cpp:619
#21 0x00005555568c87e1 in hotkey::command_executor::run_queued_commands() (this=0x555563111e40) at src/hotkey/command_executor.cpp:759
#22 0x00005555568c74d7 in hotkey::run_events(hotkey::command_executor*) (executor=0x555563111e40) at src/hotkey/command_executor.cpp:574
#23 0x00005555568c7412 in hotkey::mbutton_event(SDL_Event const&, hotkey::command_executor*) (event=..., executor=0x555563111e40) at src/hotkey/command_executor.cpp:546
#24 0x0000555556274f01 in controller_base::handle_event(SDL_Event const&) (this=0x7fffffffc600, event=...) at src/controller_base.cpp:185
#25 0x00005555568a59f9 in events::pump() () at src/events.cpp:739
#26 0x000055555627595e in controller_base::play_slice(bool) (this=0x7fffffffc600, is_delay_enabled=true) at src/controller_base.cpp:353
#27 0x0000555555df855f in playmp_controller::play_slice(bool) (this=0x7fffffffc600, is_delay_enabled=true) at src/playmp_controller.cpp:489
#28 0x0000555555deb2ec in play_controller::play_slice_catch() (this=0x7fffffffc600) at src/play_controller.cpp:1152
#29 0x0000555555df5fde in playmp_controller::play_human_turn() (this=0x7fffffffc600) at src/playmp_controller.cpp:154
#30 0x0000555555dfdc99 in playsingle_controller::play_side_impl() (this=0x7fffffffc600) at src/playsingle_controller.cpp:418
#31 0x0000555555decb3d in play_controller::play_side() (this=0x7fffffffc600) at src/play_controller.cpp:1326
#32 0x0000555555decfba in play_controller::play_turn() (this=0x7fffffffc600) at src/play_controller.cpp:1366
#33 0x0000555555dfbeeb in playsingle_controller::play_scenario_main_loop() (this=0x7fffffffc600) at src/playsingle_controller.cpp:167
#34 0x0000555555dfc906 in playsingle_controller::play_scenario(config const&) (this=0x7fffffffc600, level=...) at src/playsingle_controller.cpp:273
#35 0x0000555555b2b257 in campaign_controller::playmp_scenario(end_level_data&) (this=0x7fffffffcec0, end_level=...) at src/game_initialization/playcampaign.cpp:221
#36 0x0000555555b2b57c in campaign_controller::play_game() (this=0x7fffffffcec0) at src/game_initialization/playcampaign.cpp:284
#37 0x0000555555b34b99 in game_launcher::launch_game(game_launcher::reload_mode) (this=0x5555575b1350, reload=game_launcher::reload_mode::RELOAD_DATA) at src/game_launcher.cpp:989
#38 0x0000555555992f3a in do_gameloop(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_tr--Type <RET> for more, q to quit, c to continue without paging--
aits<char>, std::allocator<char> > > > const&) (args=std::vector of length 1, capacity 1 = {...}) at src/wesnoth.cpp:897
#39 0x0000555555993a81 in main(int, char**) (argc=1, argv=0x7fffffffe1a8) at src/wesnoth.cpp:1100
(gdb) bt
#0  0x0000555555e3cebc in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count(std::__weak_count<(__gnu_cxx::_Lock_policy)2> const&) (this=0x7fffffff97a8, __r=...)
    at /usr/include/c++/8/bits/shared_ptr_base.h:893
#1  0x0000555555f74267 in std::__shared_ptr<attack_type const, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<attack_type, void>(std::__weak_ptr<attack_type, (__gnu_cxx::_Lock_policy)2> const&)
    (this=0x7fffffff97a0, __r=...) at /usr/include/c++/8/bits/shared_ptr_base.h:1191
#2  0x0000555555f72bf7 in std::shared_ptr<attack_type const>::shared_ptr<attack_type, void>(std::weak_ptr<attack_type> const&)
    (this=0x7fffffff97a0, __r=<error reading variable: Cannot access memory at address 0x8>) at /usr/include/c++/8/bits/shared_ptr.h:276
#3  0x0000555555f71e95 in std::enable_shared_from_this<attack_type>::shared_from_this() const (this=0x0) at /usr/include/c++/8/bits/shared_ptr.h:659
#4  0x000055555601f36a in unit_display::unit_attack(display*, game_board&, map_location const&, map_location const&, int, attack_type const&, std::shared_ptr<attack_type const>, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const*, bool)
    (disp=0x7fff90edff40, board=..., a=..., b=..., damage=0, attack=..., secondary_attack=std::shared_ptr<const attack_type> (use count 3, weak count 1) = {...}, swing=1, hit_text="", drain_amount=0, att_text="", extra_hit_sounds=0x7fffffff9c10, attacking=false) at src/units/udisplay.cpp:624
#5  0x00005555561226a7 in (anonymous namespace)::attack::perform_hit(bool, statistics::attack_context&) (this=0x7fffffffa1f0, attacker_turn=false, stats=...) at src/actions/attack.cpp:1094
#6  0x0000555556125522 in (anonymous namespace)::attack::perform() (this=0x7fffffffa1f0) at src/actions/attack.cpp:1440
#7  0x00005555561263e9 in attack_unit(map_location const&, map_location const&, int, int, bool) (attacker=..., defender=..., attack_with=0, defend_with=0, update_display=true)
    at src/actions/attack.cpp:1562
#8  0x000055555612647f in attack_unit_and_advance(map_location const&, map_location const&, int, int, bool) (attacker=..., defender=..., attack_with=0, defend_with=0, update_display=true)
    at src/actions/attack.cpp:1571
#9  0x0000555555f18565 in synced_command_func_attack(config const&, bool, bool, synced_command::error_handler_function) (child=..., show=true, error_handler=...)
    at src/synced_commands.cpp:227
#10 0x0000555555f1fb2f in synced_context::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, config const&, bool, bool, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>) (commandname="attack", data=..., use_undo=true, show=true, error_handler=...) at src/synced_context.cpp:67
#11 0x0000555555f1fea5 in synced_context::run_and_store(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, config const&, bool, bool, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>) (commandname="attack", data=..., use_undo=true, show=true, error_handler=...)
    at src/synced_context.cpp:95
#12 0x0000555555f1ffae in synced_context::run_and_throw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, config const&, bool, bool, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>) (commandname="attack", data=..., use_undo=true, show=true, error_handler=...)
    at src/synced_context.cpp:109
#13 0x0000555556465edf in events::mouse_handler::attack_enemy_(map_location const&, map_location const&, int) (this=0x7fffffffc740, att_loc=..., def_loc=..., choice=0)
    at src/mouse_events.cpp:1358
#14 0x0000555556465791 in events::mouse_handler::attack_enemy(map_location const&, map_location const&, int) (this=0x7fffffffc740, attacker_loc=..., defender_loc=..., choice=0)
    at src/mouse_events.cpp:1297
#15 0x0000555556462af8 in events::mouse_handler::move_action(bool) (this=0x7fffffffc740, browse=false) at src/mouse_events.cpp:884
#16 0x0000555556462572 in events::mouse_handler::select_or_action(bool) (this=0x7fffffffc740, browse=false) at src/mouse_events.cpp:815
#17 0x0000555556431e74 in play_controller::hotkey_handler::select_and_action() (this=0x555563111e40) at src/hotkey/hotkey_handler.cpp:122
#18 0x00005555568c62d3 in hotkey::command_executor::do_execute_command(hotkey::hotkey_command const&, int, bool, bool) (this=0x555563111e40, cmd=..., press=true, release=false)
    at src/hotkey/command_executor.cpp:336
#19 0x0000555556432979 in play_controller::hotkey_handler::do_execute_command(hotkey::hotkey_command const&, int, bool, bool)
    (this=0x555563111e40, cmd=..., index=-1, press=true, release=false) at src/hotkey/hotkey_handler.cpp:275
#20 0x00005555568c7b5a in hotkey::command_executor::execute_command_wrap(hotkey::command_executor::queued_command const&) (this=0x555563111e40, command=...)
    at src/hotkey/command_executor.cpp:619
#21 0x00005555568c87e1 in hotkey::command_executor::run_queued_commands() (this=0x555563111e40) at src/hotkey/command_executor.cpp:759
#22 0x00005555568c74d7 in hotkey::run_events(hotkey::command_executor*) (executor=0x555563111e40) at src/hotkey/command_executor.cpp:574
#23 0x00005555568c7412 in hotkey::mbutton_event(SDL_Event const&, hotkey::command_executor*) (event=..., executor=0x555563111e40) at src/hotkey/command_executor.cpp:546
#24 0x0000555556274f01 in controller_base::handle_event(SDL_Event const&) (this=0x7fffffffc600, event=...) at src/controller_base.cpp:185
#25 0x00005555568a59f9 in events::pump() () at src/events.cpp:739
#26 0x000055555627595e in controller_base::play_slice(bool) (this=0x7fffffffc600, is_delay_enabled=true) at src/controller_base.cpp:353
#27 0x0000555555df855f in playmp_controller::play_slice(bool) (this=0x7fffffffc600, is_delay_enabled=true) at src/playmp_controller.cpp:489
#28 0x0000555555deb2ec in play_controller::play_slice_catch() (this=0x7fffffffc600) at src/play_controller.cpp:1152
#29 0x0000555555df5fde in playmp_controller::play_human_turn() (this=0x7fffffffc600) at src/playmp_controller.cpp:154
#30 0x0000555555dfdc99 in playsingle_controller::play_side_impl() (this=0x7fffffffc600) at src/playsingle_controller.cpp:418
#31 0x0000555555decb3d in play_controller::play_side() (this=0x7fffffffc600) at src/play_controller.cpp:1326
#32 0x0000555555decfba in play_controller::play_turn() (this=0x7fffffffc600) at src/play_controller.cpp:1366
#33 0x0000555555dfbeeb in playsingle_controller::play_scenario_main_loop() (this=0x7fffffffc600) at src/playsingle_controller.cpp:167
#34 0x0000555555dfc906 in playsingle_controller::play_scenario(config const&) (this=0x7fffffffc600, level=...) at src/playsingle_controller.cpp:273
#35 0x0000555555b2b257 in campaign_controller::playmp_scenario(end_level_data&) (this=0x7fffffffcec0, end_level=...) at src/game_initialization/playcampaign.cpp:221
#36 0x0000555555b2b57c in campaign_controller::play_game() (this=0x7fffffffcec0) at src/game_initialization/playcampaign.cpp:284
#37 0x0000555555b34b99 in game_launcher::launch_game(game_launcher::reload_mode) (this=0x5555575b1350, reload=game_launcher::reload_mode::RELOAD_DATA) at src/game_launcher.cpp:989
#38 0x0000555555992f3a in do_gameloop(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) (args=std::vector of length 1, capacity 1 = {...}) at src/wesnoth.cpp:897
#39 0x0000555555993a81 in main(int, char**) (argc=1, argv=0x7fffffffe1a8) at src/wesnoth.cpp:1100
```

**Comment by Pentarctagon** [comment]
This sounds like it might be related to #4927 ?

**Comment by mattsc** [comment]
Fixed in 7f14cc57e
